sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+            MobileCenterLog.debug(LOG_TAG, "No token, need to open browser to login url=" + url);
+
+            /* Store request id. */
+            StorageHelper.PreferencesStorage.putString(PREFERENCE_KEY_REQUEST_ID, requestId);
+
+            /* Open browser, remember that whatever the outcome to avoid opening it twice. */
+            BrowserUtils.openBrowser(url, mForegroundActivity);
+            mBrowserOpened = true;
+        }
+    }
+
+    @NonNull
+    private String computeHash(@NonNull Context context) throws PackageManager.NameNotFoundException {
+        PackageManager packageManager = context.getPackageManager();
+        PackageInfo packageInfo = packageManager.getPackageInfo(context.getPackageName(), 0);
+        return computeHash(context, packageInfo);
+    }
+
+    @NonNull
+    private String computeHash(@NonNull Context context, @NonNull PackageInfo packageInfo) {
+        // TODO switch to the following hash when backend supports it: HashUtils.sha256(context.getPackageName() + ":" + packageInfo.versionName + ":" + packageInfo.versionCode);
+        return context.getString(packageInfo.applicationInfo.labelRes);
+    }
+
+    /**
+     * Reset all variables that matter to restart checking a new release on launcher activity restart.
+     *
+     * @param releaseDetails to check if state changed and that the call should be ignored.
+     */
+    private synchronized void completeWorkflow(ReleaseDetails releaseDetails) {
+        if (releaseDetails == mReleaseDetails) {
+            completeWorkflow();
+        }
+    }
+
+    /**
+     * Reset all variables that matter to restart checking a new release on launcher activity restart.
+     *
+     * @param task to check if state changed and that the call should be ignored.
+     */
+    private synchronized void completeWorkflow(ProcessDownloadCompletionTask task) {
+        if (task == mProcessDownloadCompletionTask) {
+            completeWorkflow();
+        }
+    }
+
+    /**
+     * Reset all variables that matter to restart checking a new release on launcher activity restart.
+     */
+    private synchronized void completeWorkflow() {
+        StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_DOWNLOAD_URI);
+        mCheckReleaseApiCall = null;
+        mCheckReleaseCallId = null;
+        mUpdateDialog = null;
+        mReleaseDetails = null;
+        mWorkflowCompleted = true;
+    }
+
+    /*
+     * Store update token and possibly trigger application update check.
+     * TODO encrypt token, but where to store encryption key? If it's retrieved from server,
+     * how do we protect server call to get the key in the first place?
+     * Even having the encryption key temporarily in memory is risky as that can be heap dumped.
+     */
+    synchronized void storeUpdateToken(@NonNull String updateToken, @NonNull String requestId) {
+
+        /* Keep token for later if we are not started and enabled yet. */
+        if (mContext == null) {
```

Comment:
And I already answered before, this is not about DeepLinkActivity, this is used in download completion.